edward/models/distributions.py

```diff
@@ -182,33 +232,106 @@ def sample(self, size=1):
         """
         return self.reparam(self.sample_noise(size))
 
-    def log_prob_i(self, i, xs):
+    def log_prob(self, xs):
         """
-        log p(x_i | params)
+        log p(xs | params)
 
         Parameters
         ----------
-        i : int
-            Index of the factor to take the log density of.
-        xs : np.array
-            n_minibatch x num_vars
+        xs : tf.Tensor or np.ndarray
+            n_minibatch x self.shape
 
         Returns
         -------
-        [log p(xs[1]_i | params), ..., log p(xs[S]_i | params)]
+        tf.Tensor
+            A vector
+            [ sum_{idx in shape} log p(xs[1, idx] | params[idx]), ...,
+              sum_{idx in shape} log p(xs[S, idx] | params[idx]) ]
+        """
+        if isinstance(xs, tf.Tensor):
+            shape = get_dims(xs)
+            rank = len(shape)
+        else: # NumPy array
+            shape = xs.shape
+            rank = len(shape)
+
+        # Loop over each random variable.
+        # If univariate distribution, this is over all indices; if
+        # multivariate distribution, this is over all but the last
+        # index.
+        n_minibatch = shape[0]
+        log_prob = tf.zeros([n_minibatch], dtype=tf.float32)
+        if len(self.shape) == 1:
+            if self.is_multivariate:
+                idx = ()
+                log_prob += self.log_prob_idx(idx, xs)
+            else:
+                for i in range(self.shape[0]):
+                    idx = (i, )
+                    log_prob += self.log_prob_idx(idx, xs)
 
-        Notes
-        -----
-        This calculates the density of the ith factor, not necessarily
-        the ith latent variable (such as for multivariate factors).
+        elif len(self.shape) == 2:
+            if self.is_multivariate:
+                for i in range(self.shape[0]):
+                    idx = (i, )
+                    log_prob += self.log_prob_idx(idx, xs)
+
+            else:
+                for i in range(self.shape[0]):
+                    for j in range(self.shape[1]):
+                        idx = (i, j, )
+                        log_prob += self.log_prob_idx(idx, xs)
+
+        elif len(self.shape) == 3:
+            if self.is_multivariate:
+                for i in range(self.shape[0]):
+                    for j in range(self.shape[1]):
+                        idx = (i, j, )
+                        log_prob += self.log_prob_idx(idx, xs)
+
+            else:
+                for i in range(self.shape[0]):
+                    for j in range(self.shape[1]):
+                        for k in range(self.shape[2]):
+                            idx = (i, j, k, )
+                            log_prob += self.log_prob_idx(idx, xs)
+
+        else: # len(self.shape) >= 4
+            # There should be a generic recursive solution.
+            raise NotImplementedError()
```

Comment:
is the issue the input `idx` into `self.log_prob_idx`? if so would this work?

http://stackoverflow.com/questions/22473053/itertools-equivalent-of-nested-loop-for-x-in-xs-for-y-in-ys